azure-iot-device/tests/iothub/pipeline/test_iothub_pipeline.py

```diff
@@ -7,505 +7,1407 @@
+@pytest.mark.describe("IoTHubPipeline - EVENT: Method Request Received")
+class TestIoTHubPipelineEVENTReceiveMethodRequest(object):
+    @pytest.mark.it(
+        "Triggers the 'on_method_request_received' handler, passing the received method request as an argument"
+    )
+    def test_with_handler(self, mocker, pipeline, method_request):
+        # Set the handler
+        mock_handler = mocker.MagicMock()
+        pipeline.on_method_request_received = mock_handler
+        assert mock_handler.call_count == 0
 
-        device_pipeline_adapter.connect()
-        mock_mqtt_client_operator.on_mqtt_connected()
-        device_pipeline_adapter.disable_feature(constant.METHODS)
+        # Create the event
+        method_request_event = pipeline_events_iothub.MethodRequestEvent(method_request)
 
-        mock_mqtt_client_operator.unsubscribe.assert_called_once_with(
-            topic=subscribe_methods_topic, callback=ANY
-        )
+        # Trigger the event
+        pipeline._pipeline.on_pipeline_event(method_request_event)
+
+        assert mock_handler.call_count == 1
+        assert mock_handler.call_args == mocker.call(method_request)
+
+    @pytest.mark.it(
+        "Drops the method request if the 'on_method_request_received' handler is not set"
+    )
+    def test_no_handler(self, pipeline, method_request):
+        method_request_event = pipeline_events_iothub.MethodRequestEvent(method_request)
+        pipeline._pipeline.on_pipeline_event(method_request_event)
+
+        # No assertions required - not throwing an exception means the test passed
+
+
+@pytest.mark.describe("IoTHubPipeline - EVENT: Twin Desired Properties Patch Received")
+class TestIoTHubPipelineEVENTReceiveDesiredPropertiesPatch(object):
+    @pytest.mark.it(
+        "Triggers the 'on_twin_patch_received' handler, passing the received twin patch as an argument"
+    )
+    def test_with_handler(self, mocker, pipeline, twin_patch):
+        # Set the handler
+        mock_handler = mocker.MagicMock()
+        pipeline.on_twin_patch_received = mock_handler
+        assert mock_handler.call_count == 0
+
+        # Create the event
+        twin_patch_event = pipeline_events_iothub.TwinDesiredPropertiesPatchEvent(twin_patch)
+
+        # Trigger the event
+        pipeline._pipeline.on_pipeline_event(twin_patch_event)
+
+        assert mock_handler.call_count == 1
+        assert mock_handler.call_args == mocker.call(twin_patch)
+
+    @pytest.mark.it("Drops the twin patch if the 'on_twin_patch_received' handler is not set")
+    def test_no_handler(self, pipeline, twin_patch):
+        twin_patch_event = pipeline_events_iothub.TwinDesiredPropertiesPatchEvent(twin_patch)
+        pipeline._pipeline.on_pipeline_event(twin_patch_event)
+
+        # No assertions required - not throwing an exception means the test passed
+
+
+# connection_string_format = "HostName={};DeviceId={};SharedAccessKey={}"
```

Comment:
>connection_string_format [](start = 2, length = 24)

any reason these commented lines are around

Reply:
whoops

---
In reply to: [299254690](https://github.com/Azure/azure-iot-sdk-python-preview/pull/120#discussion_r299254690) [](ancestors = 299254690)